crates/bevy_hierarchy/src/child_builder.rs

```diff
@@ -96,7 +93,8 @@ fn update_old_parent(world: &mut World, child: Entity, parent: Entity) {
 ///
 /// Sends [`HierarchyEvent`]'s.
 fn update_old_parents(world: &mut World, parent: Entity, children: &[Entity]) {
-    let mut events: SmallVec<[HierarchyEvent; 8]> = SmallVec::with_capacity(children.len());
+    let mut events: SmallVec<[(OnParentChange, Entity); 8]> =
```

Comment:
Yeah I don't think we need deferred execution here, so I just removed the event buffers and `trigger_events` in favor of simply calling `world.trigger_targets(...)` directly.

For `remove_children` I had to clone the `SmallVec` contained in `Children` to get around issues with simultaneous mutable and immutable `World` access. I guess maybe there could be some unsafe way to avoid cloning (or some other approach I'm missing), but then you'd need to ensure that `parent != child`.

Reply:
Regarding eliminating the `clone()` in `remove_children`, I have two ideas:

i) You could move the call to `world.get()` inside the loop. This means that `&Children` is only borrowed from `world` until the call to `parent_children.contain()` finishes, and then you're free to mutably borrow `world` in order to trigger the observer, etc. The downside is that when removing multiple children you have to pay the overhead of looking up the entity and component multiple times rather than once.

```rust
for child in children {
    let Some(parent_children) = world.get::<Children>(parent) else {
        return;
    };
    if parent_children.contains(child) {
        // ...
```

ii) You could temporarily remove the vector of parent's children from the world and put it back afterwards. This can be efficiently accomplished by swapping the `Children` component with an empty one. If the vector is large enough to have required the `SmallVec` to allocate then this is just a pointer swap. The downside is that the parent's `Children` component will appear empty to any observers examining the world during the remove operation.

This approach also lets you combine processing the child entities with removing elements from the `parent_children` vector, and hence halves the number of calls to `contains()`.

```rust
fn remove_children(parent: Entity, children: &[Entity], world: &mut World) {
    let Some(mut borrowed_parent_children) = world.get_mut::<Children>(parent) else {
        return;
    };

    // Temporarily take the vector of parent's children out of the world.
    let mut parent_children = Children::from_entities(&[]);
    mem::swap(&mut parent_children, borrowed_parent_children.as_mut());

    // Remove children and trigger observers
    parent_children.0.retain(|child| {
        let remove = children.contains(child);
        if remove {
            world.trigger_targets(OnParentChange::Removed(parent), *child);
            world.entity_mut(*child).remove::<Parent>();
        }
        !remove
    });

    // Return the parent's children to the world if there are any left.
    let mut parent = world.entity_mut(parent);
    if parent_children.is_empty() {
        parent.remove::<Children>();
    } else {
        if let Some(mut borrowed_parent_children) = parent.get_mut::<Children>() {
            mem::swap(&mut parent_children, borrowed_parent_children.as_mut());
        }
    }
}
```

Since this code involves triggering an observer in the loop, I don't think there any is way you could rely on unsafe primitives to help as an observer could break any invariants.

Anyway, I'm really just nitpicking :-). I think this is a good use case for observers.